satpy/composites/__init__.py

```diff
@@ -1039,15 +1059,26 @@ def __init__(self, name, filename=None, url=None, known_hash=None, area=None,
 
     @staticmethod
     def _get_cache_filename_and_url(filename, url):
-        if filename is not None:
-            filename = os.path.expanduser(os.path.expandvars(filename))
-        if url is not None:
+        if url:
             url = os.path.expandvars(url)
-            if filename is None:
+            if not filename:
                 filename = os.path.basename(url)
```

Comment:
By switching the order that url and filename are checked, I think you've made it so `filename` doesn't have environment variables expanded when `url` is provided. That's not a huge issue, but I think it is an unnecessary limitation here. I think the code can be cleaner if you swap them back (check filename first, then check URL). You'd just need to change `if not os.path.isabs(filename):` to `if not os.path.isabs(filename) and not url:`. Thoughts?

Reply:
I did not realized the use case of expandable filename passed to URL. I have switched code back. Thanks.

Reply:
It is an edge case, but not something I see as impossible.